### PR TITLE
Improved interpolations

### PR DESCRIPTION
**Pull request by yakir12**
This fixes the interpolation problems mentioned in #9. 

The core of the problem is the introduction of zeros along the perimeter of the bounding geometry. Instead, the 4 corners of the bounding rectangle that contains all of the data (including the geometry) are added and set to `pad_value` (e.g. zero). When the geometry is a circle, this solves cases where a known data point is larger than zero AND close to the perimeter of the geometry. Because I refrain from setting the perimeter data to zero, using the `DelaunayMesh` is completely impossible. We would need to be able to interpolate such a mesh for that to work. Ideally, https://github.com/JuliaMath/Interpolations.jl/issues/118 
would be solved and we could simply use Julia's `Interpolations.jl`. 

This PR is quite aggressive, so I expect some pushback and appreciate any feedback you may have.

### CODE REVIEW

**Comment by yakir12** [comment]
I'll go ahead and ping @behinger in case he's not on parental leave yet 😬 and @palday, maybe you could review this PR. Thanks!

File: src/core-recipe.jl [hunk 1]
```diff
@@ -137,29 +157,3 @@ function enclosing_geometry(::Type{Rect}, positions, enlarge=0.0)
...
-function pad_data(data::AbstractVector, npositions::Integer, value::Number)
-    ndata = length(data)
-    if npositions == ndata
-        return data
-    elseif npositions < ndata
-        error("To pad the data for new positions, we need more positions than data points")
-    else
-        vcat(data, fill(value, npositions - ndata))
-    end
-end
```
Comment (palday) [inline src/core-recipe.jl:165]: let's add these back in since they're causing the docs build to fail and #12 should actually use them

File: src/interpolators.jl [hunk 2]
```diff
@@ -80,6 +81,7 @@ function delaunay_mesh(positions::AbstractVector{<: Point{2}})
     m = delaunay(convert(Matrix{Float64}, hcat(first.(positions), last.(positions))))
     return GeometryBasics.Mesh(Makie.to_vertices(m.points), Makie.to_triangles(m.simplices))
 end
+=#
```
Comment (palday) [inline src/interpolators.jl:84]: I would do line comments for each impacted line instead of a block comment -- it makes it much more obvious when looking at a given line that it's been commented out

File: test/runtests.jl [hunk 3]
```diff
@@ -24,24 +24,6 @@ begin
...
-        f[2, 1],
-        data_obs, positions,
-        interpolation=DelaunayMesh(),
-        labels = string.(1:length(positions)),
-        colorrange=(-1, 1),
-        colormap=[:red, :blue],
-        axis=(title="delaunay mesh", aspect=DataAspect(),))
-    display(f)
-    @test_figure("delaunay-with-slider", f)
-end
```
Comment (palday) [inline test/runtests.jl:43]: can you change this to "just" commented out until we actually delete the Delaunay code?

File: src/core-recipe.jl [hunk 4]
```diff
@@ -62,17 +62,37 @@ macro plot_or_defaults(var, defaults)
...
+    xmin, ymin = minimum(circ)
+    xmax, ymax = maximum(circ)
+    Rect2f(xmin, ymin, xmax - xmin, ymax - ymin)
+end
+
+function get_bounded(rect, positions, data, padding, pad_value)
+    bb_xy = decompose(Point2f, rect)
+    bb_v = fill(pad_value, 4)
+    @assert padding ≥ 0 "cannot deal with negative padding"
+    @assert allunique(positions) "positions must be unique"
```
Comment (palday) [inline src/core-recipe.jl:77]:
```suggestion
    padding ≥ 0 || error("cannot deal with negative padding")
    allunique(positions) || error("positions must be unique")
```

`@assert` can be optimized out in some situations

File: src/core-recipe.jl [hunk 5]
```diff
@@ -86,22 +106,22 @@ function Makie.plot!(p::TopoPlot)
...
+        if geometry isa Circle
+            c = geometry.center
+            r = geometry.r
+            out = [norm(Point2f(x, y) - c) > r for x in xg, y in yg]
+            z[out] .= NaN
         end
+        return z
+    end
+
+    heatmap!(p, xg, yg, data, colormap=p.colormap, colorrange=p.colorrange, interpolate=true)
```
Comment (palday) [inline src/core-recipe.jl:120]: @SimonDanisch thoughts/feelings on using a heatmap instead of a mesh?

File: src/interpolators.jl (hunk 2 above)
Comment (yakir12) [reply src/interpolators.jl:84]: I used block comments cause that is what was used to comment out the badly behaving MNE interpolation, so I used the same reasoning when commenting out the badly behaved Delaunay interpolation. But I changed that now to line commenting.

File: src/core-recipe.jl (hunk 1 above)
Comment (yakir12) [reply src/core-recipe.jl:165]: I disagree. `pad_data` and `pad_boundary!` work only by padding the data with `pad_value` along the perimeter, which is what caused issue #9. They are never used in this PR. Instead of adding those back, I removed their links in the documentations, which fixes the problem with building the docs -- locally (see if it works in CI).

File: src/core-recipe.jl (hunk 5 above)
Comment (yakir12) [reply src/core-recipe.jl:120]: We can use either:
1. `heatmap` uses the same gridded data we generate for the `contour` plot. Reusing the same data is cheap and also means that the contour lines and the heatmap's colors match perfectly (cause it's literally based on the same data).
2. `mesh` requires Delaunay triangulation to find the coordinates that best work with the a) original coordinates and b) the coordinates of the perimeter (i.e. `geometry`). We then need to interpolate the data for the coordinates of the triangles in that Delaunay mesh. There will probably be a slight mismatch between the resulting contour lines and color gradient, but I doubt it'll be noticeable?

The advantage of a `mesh` over a `heatmap` is that the edges of a mesh are "vectorized" while the colormap will be pixelated. So if you zoom in on the edge of a circle, you'll see the individual pixels of the heatmap (and the will depend on `resolution`), while in a mesh it'll just be a line (connecting the number of angles you used to decompose said circle, e.g. 64).

Comment (palday) [reply src/core-recipe.jl:122]: I think a `heatmap` is fine for the moment, but we can always revisit this decision later.

**Comment by palday** [comment]
@yakir12 can you add in a test for peaks not shifting away from their nominal location? Te example using four points near the boundary was a good one.

**Comment by yakir12** [comment]
I'll keep this open in case you need it for now.

**Comment by behinger** [comment]
pad_value prior could be an integer, now has to be a float. Not sure where this cropped up. Not a biggy, but took me a bit to debug

the results look much better
old

new



but, the interpolations still look quite different (albeit much better), compared to PyMNE


as a more comparable comparison (imho), this is with extrapolate="box"


MNE Plot via:
```
posMat = Matrix(hcat(pos...)).-0.5
# you can also make .*0.5 afterwards, then all points are within the head
	#pos = PyMNE.channels.make_eeg_layout(mon).pos
PyMNE.viz.plot_topomap(rand(MersenneTwister(1),length(pos)),posMat',cmap="RdBu_r")
```

**Comment by behinger** [comment]
another point to discuss: does this work also for other bounding geometries? e.g. if you have a weird polynomial, you might want to deactivate this functionality? Or should this even be "off" by default, and eeg_topoplot activates it?

**Comment by palday** [comment]
Superseded by #20
